### PR TITLE
feat: replace bloom with xor8 filter

### LINKED ISSUE #510: Xor Filters: Faster and Smaller Than Bloom Filters
### Describe This Problem

In currently implementation, bloom filters are used to reduce unnecessary data involved in one Query. There is one annoying issues:
- large memory consumption, every bloom filter is 256 bytes.

### Proposal

There is an interesting algorithm Xor filters to address bloom filters' issue.

image

### Additional Context

- https://lemire.me/blog/2019/12/19/xor-filters-faster-and-smaller-than-bloom-filters/
- https://bohutang.me/2022/11/21/databend-xor-filter/

### PR DESCRIPTION
**Pull request by jiacai2050**
# Which issue does this PR close?

Closes #510

# Rationale for this change
 
See issue above

# What changes are included in this PR?

- Replace bloom with xor8 filter
- Add `Filter` trait to represent different filters

# Are there any user-facing changes?

No

# How does this change test

Old UT

### CODE REVIEW

File: analytic_engine/src/sst/parquet/meta_data.rs [hunk 1]
```diff
@@ -45,89 +58,201 @@ pub enum Error {
...
         }
     }
 
-    /// Accrue the data belonging to one column.
-    ///
-    /// Caller should ensure the `column_idx` is in the range.
-    pub fn accrue_column_data(&mut self, column_idx: usize, data: &[u8]) {
-        if self.column_filters[column_idx].is_none() {
-            self.column_filters[column_idx] = Some(Bloom::default());
+    pub(crate) fn add_key(&mut self, col_idx: usize, key: &[u8]) {
```
Comment (Rachelint) [inline analytic_engine/src/sst/parquet/meta_data.rs:121]: How about defining a trait `FilterBuilder`?

File: analytic_engine/src/sst/parquet/meta_data.rs [hunk 2]
```diff
@@ -45,89 +58,201 @@ pub enum Error {
...
+    xor8: Xor8,
+}
+
+impl fmt::Debug for Xor8Filter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("XorFilter")
+    }
+}
+
+impl Filter for Xor8Filter {
```
Comment (Rachelint) [inline analytic_engine/src/sst/parquet/meta_data.rs:91]: How about `ColumnFilter`?

File: analytic_engine/src/sst/parquet/row_group_pruner.rs [hunk 3]
```diff
@@ -16,101 +16,101 @@ use parquet_ext::prune::{
...
+                // TODO: We can do continuous prune based on the `pruned0` to reduce the
                 // filtering cost.
-                let filtered1 = self.filter_by_bloom(v);
-                let filtered = Self::intersect_filtered_row_groups(&filtered0, &filtered1);
+                let pruned1 = self.prune_by_filters(v);
+                let pruned = Self::intersect_pruned_row_groups(&pruned0, &pruned1);
 
                 debug!(
-                    "Finish filtering row groups by blooms and min_max, total_row_groups:{}, filtered_by_min_max:{}, filtered_by_blooms:{}, filtered_by_both:{}",
+                    "Finish prune row groups by sst_filter and min_max, total_row_groups:{}, pruned_by_min_max:{}, pruned_by_blooms:{}, pruned_by_both:{}",
```
Comment (Rachelint) [inline analytic_engine/src/sst/parquet/row_group_pruner.rs:73]: Finish pruning

File: components/parquet_ext/src/prune/equal.rs [hunk 4]
```diff
@@ -478,7 +478,7 @@ mod tests {
             Field::new("c2", DataType::Int32, false),
         ]);
         let filtered_row_groups =
```
Comment (Rachelint) [inline components/parquet_ext/src/prune/equal.rs:480]: target_row_groups?

File: analytic_engine/src/sst/parquet/meta_data.rs (hunk 2 above)
Comment (jiacai2050) [reply analytic_engine/src/sst/parquet/meta_data.rs:91]: It's a general filter, not designed for column, so I prefer current name.

File: analytic_engine/src/sst/parquet/meta_data.rs (hunk 1 above)
Comment (jiacai2050) [reply analytic_engine/src/sst/parquet/meta_data.rs:121]: Good idea, I plan to move Filter related code to `src/meta_data.rs`, currently it's in wrong place.

I left a comment to add FilterBuilder trait when refactor this.

File: analytic_engine/src/sst/parquet/meta_data.rs [hunk 5]
```diff
@@ -45,89 +58,201 @@ pub enum Error {
...
+            .as_ref()
+            .map(|v| v.contains(data))
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
-pub struct BloomFilter {
+// TODO: move this to sst module
+#[derive(Debug, Clone, PartialEq, Default)]
+pub struct SstFilter {
```
[the remark this answers is not in the document]
Comment (jiacai2050) [reply analytic_engine/src/sst/parquet/meta_data.rs:195]: OK, this make sense to me.

**Review by Rachelint — APPROVED** [review]
LGTM

**Pull request merged**